modules/drivers/sqlite/src/metabase/driver/sqlite.clj

```diff
@@ -360,6 +361,97 @@
   [_]
   (hsql/call :datetime (hx/literal :now)))
 
+(defmethod sql.qp/->honeysql [:sqlite :datetime-diff]
+  [driver [_ x y unit]]
+  (let [x                (sql.qp/->honeysql driver x)
+        y                (sql.qp/->honeysql driver y)
+        extract          (fn [unit x] (sql.qp/date :sqlite unit x))
+        disallowed-types (keep
+                          (fn [v]
+                            (some->> v
+                                     hx/type-info
+                                     hx/type-info->db-type
+                                     name
+                                     (re-find #"(?i)^time")))
+                          [x y])]
+    (when (seq disallowed-types)
+      (throw (ex-info (tru "Only datetime, timestamp, or date types allowed. Found {0}"
+                           (pr-str disallowed-types))
+                      {:found disallowed-types
+                       :type  qp.error-type/invalid-query})))
+    (case unit
+      :year
+      (let [positive-diff
+            (fn [a b] ; precondition: a <= b
+              (hx/-
+               (hx/- (extract :year b) (extract :year a))
+               ;; decrement if a is later than b in the year calendar
+               (hx/cast
+                :integer
+                (hsql/call
+                 :or
+                 (hsql/call :> (extract :month-of-year a) (extract :month-of-year b))
+                 (hsql/call
+                  :and
+                  (hsql/call := (extract :month-of-year a) (extract :month-of-year b))
+                  (hsql/call :> (extract :day-of-month a) (extract :day-of-month b)))))))]
+        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x))))
+
+      :quarter
+      (let [positive-diff
+            (fn [a b]
+              (hx/cast
+               :integer
+               (hx/floor (hx// (hx/- (hx/+ (hx/* (hx/- (extract :year b)
+                                                       (extract :year a))
+                                                 12)
+                                           (hx/- (extract :month-of-year b)
+                                                 (extract :month-of-year a)))
+                                     (hx/cast
+                                      :integer
+                                      (hsql/call :> (extract :day-of-month a) (extract :day-of-month b))))
+                               3))))]
+        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x))))
```

Comment:
Any time you're doing crazy math like this in the HoneySQL you generate, you should at least write a comment explaining what in the world is going on. Otherwise trying to visualize what this does will make people's eyes bleed. 

Take a look at this code in the H2 driver for example

https://github.com/metabase/metabase/blob/43de02b17c73604d503a7ec501d2491dd867d030/src/metabase/driver/h2.clj#L193-L207

It does something that is similarly complicated but does a good job of explaining why it's doing it and what it's doing

The rule of thumb to use here is that if it's not immediately obvious what the code is doing and why from reading it, it's probably a good idea to write a quick comment explaining what's happening. Just seeing an example of the SQL we're trying to generate would be **enormously** helpful here.